#### 1. Запропонуваємо розташування засобів захисту для мережі, об’єкти захисту якої визначені 10-им варіантом ↓

<img src="https://github.com/sazan24/KPI/blob/main/Bachelor's%20degree/4%20course/Integrated%20Information%20Security%20Systems/Scheme%20of%20network%20connections.png?raw=true" alt="Scheme of network connections">

#### 2. Сформулюємо проблему (2.1), (2.2) у виді задачі оптимізації із зазначенням семантики постановки задачі ↓

##### • 1) **Визначаємо елементи**, між якими потрібно знайти компроміс:

Нехай: $x_1$ - кількість апаратних IDS/IPS у заданій системі, $x_2$ – кількість ММЕ, $x_3$ - кількість маршрутизаторів із функціям ММЕ та IDS/IPS.

##### • 2) Нехай етапи, для яких буде визначатись **питання компромісу**, виглядадимуть наступним чином:

– Закупівля засобів захисту

– Встановлення засобів захисту

– Розгортання системи

– Адміністрування та підтримка системи

##### • 3) Визначимо (**вводимо на власний розсуд**) вартості засобів захисту на різних етапах життя системи:

Отримуємо відповідну матрицю вартості реалізації процесів для кожного із 
компонентів системи захисту:

C = 
$\begin{pmatrix}
2 & 5 & 7\\
3 & 1 & 3\\
2 & 2 & 5\\
4 & 1 & 6
\end{pmatrix}$

##### • 4) Визначимо обмеження на затрати на кожному етапі (знову ж таки, **вводимо на власний розсуд**):

$c_1^{max} = 16; c_2^{max} = 8; c_3^{max} = 10; c_4^{max} = 14$


##### • 5) Визначимо обмеження на затрати на кожному етапі (знову ж таки, **вводимо на власний розсуд**):

$s_1 = 0.4; s_2 = 0.3; s_3 = 0.3$

##### • 6) Нарешті, отримуємо **задачу лінійного програмування** у вигляді формул 2.1 та 2.2:

Обмеження на загальну суму витрат по кожному процесу:

► $2x_1 + 5x_2 + 7x_3 \le 16$

► $3x_1 + 5x_2 + 6x_3 \le 8$

► $5x_1 + 2x_2 +7x_3 \le 10$

► $4x_1 + x_2 + 8x_3 \le 14$

Максимізована функція:

○ $f(\vec{x}) = 0.4x_1 +0.3x_2 + 0.3x_3 → max$

##### • 7) Вирішуємо задачу за допомогою **симплекс-методу**, використовуючи при цьому онлайн-калькулятор:

Переглянемо наступний результат обчислень за посиланням, після чого запишемо у наступнній клітинці загальний висновок: https://linprog.com/en/main-simplex-method/result;queryParams=%7B%22n%22:4,%22m%22:3,%22max_min%22:1,%22values%22:%5B%5B%222%22,%225%22,%227%22,%2216%22%5D,%5B%223%22,%221%22,%223%22,%228%22%5D,%5B%222%22,%222%22,%225%22,%2210%22%5D,%5B%224%22,%221%22,%226%22,%2214%22%5D%5D,%22function%22:%5B%220.4%22,%220.3%22,%2203%22%5D,%22equalSign%22:%5B%222%22,%222%22,%222%22,%222%22%5D%7D

**Оптимальне значення функції $f = 6$ досягається у точці з координатами $(0, 0, 2)$**. Відповідно, буде зрозуміло, скільки яких засобів варто використовувати для КСЗІ. У даному випадку є сенс придбати 2 маршрутизатори із підтримкою ММЕ та IDS/IPS, які будуть виконувати ті самі функції, що й окремо ММЕ та IDS/IPS.

#### 3. Побудуємо матрицю Сааті та виконаємо відповідні обчислення вагових коефіцієнтів ↓

##### 3.1. Заповнимо (експертним методом) матрицю Сааті для порівняння важливості вимог: «Швидкодія доступу до ДМЗ», «Захист від НСД внутрішньої мережі», «Надійність роботи ДМЗ», «Витрати на обладнання», «Легкість адміністрування».

<table style="width:100%">
  <tr>
    <th style="text-align:left"><u>---|Показники|---</u></th>
    <th style="text-align:center">Швидкодія</th>
    <th style="text-align:center">Захист</th>
    <th style="text-align:center">Надійність</th>
    <th style="text-align:center">Витрати</th>
    <th style="text-align:center">Легкість</th>
  </tr>
  <tr>
    <td><strong>Швидкодія доступу до ДМЗ</strong></td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">7</td>
    <td style="text-align:center">9</td>
  </tr>
  <tr>
    <td><strong>Захист від НСД внутрішньої мережі</strong></td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
  </tr>
  <tr>
    <td><strong>Надійність роботи ДМЗ</strong></td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
  </tr>
  <tr>
    <td><strong>Витрати на обладнання</strong></td>
    <td style="text-align:center">1/7</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">1/3</td>
  </tr>
  <tr>
    <td><strong>Легкість адміністрування</strong></td>
    <td style="text-align:center">1/9</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">1</td>
  </tr>
</table>

##### 3.2. Обрахуємо вагові коефіцієнти 1-им способом, де необхідно підраховувати суми по рядках, далі суми нормуються, і, таким чином, одержимо ваги $w_i$.

In [1]:
import numpy as np

# Матриця попарних порівнянь Сааті
A = np.array([[1, 3, 5, 7, 9],
              [1/3, 1, 3, 5, 3],
              [1/5, 1/3, 1, 5, 3],
              [1/7, 1/5, 1/5, 1, 1/3],
              [1/9, 1/3, 1/3, 3, 1]])

# Підрахуємо суми по кожному рядку матриці A
row_sums = A.sum(axis=1)
print(f"Загальна сума: {sum(row_sums)}")

# Обчислимо ваги для кожного рядка
weights = row_sums / row_sums.sum()
print(f"Сума всіх ваг: {sum(weights)}")

# Виведемо значення "Сума" та "Вага" для кожного рядка
for i, weight in enumerate(weights):
    print(f"Рядок {i+1}: Сума = {row_sums[i]:.2f}, Вага = {weight:.2f}")

Загальна сума: 53.52063492063491
Сума всіх ваг: 1.0
Рядок 1: Сума = 25.00, Вага = 0.47
Рядок 2: Сума = 12.33, Вага = 0.23
Рядок 3: Сума = 9.53, Вага = 0.18
Рядок 4: Сума = 1.88, Вага = 0.04
Рядок 5: Сума = 4.78, Вага = 0.09


<table style="width:100%">
  <tr>
    <th style="text-align:left"><u>---|Показники|---</u></th>
    <th style="text-align:center">Швидкодія</th>
    <th style="text-align:center">Захист</th>
    <th style="text-align:center">Надійність</th>
    <th style="text-align:center">Витрати</th>
    <th style="text-align:center">Легкість</th>
    <th style="text-align:center">Сума</th>
    <th style="text-align:center">Вага</th>
  </tr>
  <tr>
    <td><strong>Швидкодія доступу до ДМЗ</strong></td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">7</td>
    <td style="text-align:center">9</td>
    <td style="text-align:center">25</td>
    <td style="text-align:center">0.47</td>
  </tr>
  <tr>
    <td><strong>Захист від НСД внутрішньої мережі</strong></td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">12.33</td>
    <td style="text-align:center">0.23</td>
  </tr>
  <tr>
    <td><strong>Надійність роботи ДМЗ</strong></td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">9.53</td>
    <td style="text-align:center">0.18</td>
  </tr>
  <tr>
    <td><strong>Витрати на обладнання</strong></td>
    <td style="text-align:center">1/7</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1.88</td>
    <td style="text-align:center">0.04</td>
  </tr>
  <tr>
    <td><strong>Легкість адміністрування</strong></td>
    <td style="text-align:center">1/9</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">4.78</td>
    <td style="text-align:center">0.09</td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td style="text-align:center"><strong>Усього:</strong></td>
    <td style="text-align:center"><strong>53,52</strong></td>
    <td style="text-align:center"><strong>1,00</strong></td>
  </tr>
</table>

##### 3.3. Обрахуємо вагові коефіцієнти 1-им способом, де необхідно розв’язати задачу знаходження власних значень матриці Сааті: **(А-λІ)х = 0**.

In [2]:
# Знаходимо власні значення та власні вектори
eigenvalues, eigenvectors = np.linalg.eig(A)

# Знаходимо індекс максимального власного значення
max_eigenvalue_index = np.argmax(eigenvalues)

# Вибираємо відповідний власний вектор
max_eigenvector = eigenvectors[:, max_eigenvalue_index]

# Нормалізуємо власний вектор
normalized_max_eigenvector = max_eigenvector / np.sum(max_eigenvector)

# Ваги показників
weights = normalized_max_eigenvector

# Виведемо ваги кожного показника
for i, weight in enumerate(weights):
    print(f"Вага для показника №{i + 1}: {weight:.2f}")
    
print(f"Сума усіх ваг: {np.sum(weights):.2f}")

Вага для показника №1: 0.53-0.00j
Вага для показника №2: 0.23-0.00j
Вага для показника №3: 0.14-0.00j
Вага для показника №4: 0.04-0.00j
Вага для показника №5: 0.07-0.00j
Сума усіх ваг: 1.00+0.00j


<table style="width:100%">
  <tr>
    <th style="text-align:left"><u>---|Показники|---</u></th>
    <th style="text-align:center">Швидкодія</th>
    <th style="text-align:center">Захист</th>
    <th style="text-align:center">Надійність</th>
    <th style="text-align:center">Витрати</th>
    <th style="text-align:center">Легкість</th>
    <th style="text-align:center">Вага</th>
  </tr>
  <tr>
    <td><strong>Швидкодія доступу до ДМЗ</strong></td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">7</td>
    <td style="text-align:center">9</td>
    <td style="text-align:center">0.53</td>
  </tr>
  <tr>
    <td><strong>Захист від НСД внутрішньої мережі</strong></td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">0.23</td>
  </tr>
  <tr>
    <td><strong>Надійність роботи ДМЗ</strong></td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">5</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">0.14</td>
  </tr>
  <tr>
    <td><strong>Витрати на обладнання</strong></td>
    <td style="text-align:center">1/7</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1/5</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">0.04</td>
  </tr>
  <tr>
    <td><strong>Легкість адміністрування</strong></td>
    <td style="text-align:center">1/9</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">1/3</td>
    <td style="text-align:center">3</td>
    <td style="text-align:center">1</td>
    <td style="text-align:center">0.07</td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td style="text-align:center"><strong>Усього:</strong></td>
    <td style="text-align:center"><strong>1,00</strong></td>
  </tr>
</table>